### PR TITLE
Restore snapshot from onboarding

### LINKED ISSUE #3910: Add UI to upload snapshots before onboarding
## The request

Before the onboarding starts (where you create your first user), there should be a link/button that will allow you to use a snapshot to set up Home Assistant.

image

## The alternatives

There are several alternatives to this route:

- Add upload to the snapshots tab inside the supervisor.
- Continue as now (finish onboarding, install an add-on, use the add-on to move the snapshot, restore)
- Have the supervisor pick up and apply the snapshot the same way as network config (form USB)

## Additional information

This feature request has been raised on several occasions, in various forms:
- https://github.com/home-assistant/home-assistant-polymer/issues/1356
- https://github.com/home-assistant/home-assistant-polymer/issues/1441
- https://github.com/home-assistant/home-assistant-polymer/issues/1710

This is not a simple task to "just implement", this will require work in both the [supervisor](https://github.com/home-assistant/supervisor) and the [frontend](https://github.com/home-assistant/frontend) and potentially also the [core](https://github.com/home-assistant/core)

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change

restore

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
enable_feature:
  upload_snapshot: True
```

## Additional information

- This PR fixes or closes issue: closes #3910
- This PR is related to issue:
- Link to documentation pull request:
- needs: 
  - https://github.com/home-assistant/core/pull/40585
  - https://github.com/home-assistant/core/pull/40582
  - https://github.com/home-assistant/core/pull/40617

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by aidbish** [comment]
@ludeeus.  Some folks are gonna love you for this

File: hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts [hunk 1]
```diff
@@ -30,7 +32,11 @@ export class DialogHassioSnapshotUpload extends LitElement
   }
 
   public closeDialog(): void {
-    this._params?.reloadSnapshot();
+    if (!this._params?.onboarding) {
+      if (this._params?.reloadSnapshot) {
+        this._params.reloadSnapshot();
+      }
+    }
```
Comment (bramkragten) [inline hassio/src/dialogs/snapshot/dialog-hassio-snapshot-upload.ts:39]:
```suggestion
    if (this._params && !this._params.onboarding) {
      if (this._params.reloadSnapshot) {
        this._params.reloadSnapshot();
      }
    }
```

File: src/onboarding/onboarding-restore-snapshot.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  @internalProperty() private _log?: string;
+
+  protected render(): TemplateResult {
+    return this.restoring
+      ? html`<ha-card>
+          <h2>
+            ${this.localize("ui.panel.page-onboarding.restore.in_progress")}
+          </h2>
+          ${this._log
```
Comment (bramkragten) [inline src/onboarding/onboarding-restore-snapshot.ts:43]: Let's not show the log by default, but add a button to show it.

File: src/onboarding/onboarding-restore-snapshot.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+  private async _getLogs(): Promise<void> {
+    if (this.restoring) {
+      try {
+        const response = await fetch("/api/hassio/supervisor/logs", {
+          method: "GET",
+        });
+        const logs = await response.text();
+        this._log = this._filterLogs(logs);
+        if (this._log.match(/\d{2}:\d{2}:\d{2}\s.*Restore\s\w+\sdone/)) {
```
Comment (bramkragten) [inline src/onboarding/onboarding-restore-snapshot.ts:94]: Could we do this differently? Seems like a hack to check the logs for a text

File: src/onboarding/onboarding-restore-snapshot.ts (hunk 2 above)
Comment (ludeeus) [reply src/onboarding/onboarding-restore-snapshot.ts:43]: I think the majority will want to see this log during the restore, without it just looks like it just sits there doing nothing.

File: src/onboarding/onboarding-restore-snapshot.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+          25% {
+            content: ".";
+          }
+          50% {
+            content: "..";
+          }
+          75% {
+            content: "...";
+          }
+        }
```
Comment (bramkragten) [inline src/onboarding/onboarding-restore-snapshot.ts:143]: Use a spinner?

File: src/onboarding/onboarding-restore-snapshot.ts (hunk 3 above)
Comment (ludeeus) [reply src/onboarding/onboarding-restore-snapshot.ts:94]: Not currently, but when we have <https://github.com/home-assistant/supervisor/issues/1946> we might

File: src/onboarding/onboarding-restore-snapshot.ts (hunk 4 above)
Comment (ludeeus) [reply src/onboarding/onboarding-restore-snapshot.ts:143]: Had a spinner initially, but this looked cleaner I think?

**Comment by bramkragten** [comment]
- This isn't a partial restore is it? The message in the dialog says that?
- The restore log card needs some padding

File: src/onboarding/onboarding-restore-snapshot.ts (hunk 2 above)
Comment (bramkragten) [reply src/onboarding/onboarding-restore-snapshot.ts:43]: Ideally we would have 1 line stating what it is doing right now, in not too technical words with a spinner, then a details buttons that shows the complete log.

**Comment by ludeeus** [comment]
The example(gif) was a partial snapshot

File: src/onboarding/onboarding-restore-snapshot.ts (hunk 2 above)
Comment (ludeeus) [reply src/onboarding/onboarding-restore-snapshot.ts:43]: So just the last line of the log without the time?

Comment (bramkragten) [reply src/onboarding/onboarding-restore-snapshot.ts:43]: Yeah I think so

File: src/onboarding/onboarding-restore-snapshot.ts (hunk 4 above)
Comment (bramkragten) [reply src/onboarding/onboarding-restore-snapshot.ts:143]: We use spinners everywhere, we could use `<mwc-linear-progress indeterminate></mwc-linear-progress>` If you prefer that

Comment (ludeeus) [reply src/onboarding/onboarding-restore-snapshot.ts:143]: With the last log entry, a regular spinner makes sense when the full log is not shown

File: src/onboarding/ha-onboarding.ts [hunk 5]
```diff
@@ -127,6 +144,28 @@ class HaOnboarding extends litLocalizeLiteMixin(HassElement) {
...
+        "Home Assistant Supervised",
+      ].includes(response.installation_type);
+      if (this._supervisor) {
+        // Only load if we have supervisor
+        import(
+          /* webpackChunkName: "onboarding-restore-snapshot" */ "./onboarding-restore-snapshot"
+        );
+      }
+    } catch (err) {
+      alert("Something went wrong loading onboarding, try refreshing");
```
Comment (bramkragten) [inline src/onboarding/ha-onboarding.ts:165]: We can just log this? Only thing that would not work is restoring a snapshot?

**Review by bramkragten — APPROVED** [review]

File: src/onboarding/ha-onboarding.ts (hunk 5 above)
Comment (ludeeus) [reply src/onboarding/ha-onboarding.ts:165]: True

**Pull request merged**